src/lightning/pytorch/trainer/trainer.py

```diff
@@ -1068,6 +1071,41 @@ def __setup_profiler(self) -> None:
         self.profiler._lightning_module = proxy(self.lightning_module)
         self.profiler.setup(stage=self.state.fn, local_rank=local_rank, log_dir=self.log_dir)
 
+    @contextmanager
+    def init_module(self, empty_init: Optional[bool] = None) -> Generator:
+        """Tensors that you instantiate under this context manager will be created on the device right away and
+        have the right data type depending on the precision setting in the Trainer.
+
+        The parameters and tensors get created on the device and with the right data type right away without wasting
+        memory being allocated unnecessarily. The automatic device placement under this context manager is only
+        supported with PyTorch 2.0 and newer.
+
+        Args:
+            empty_init: Whether to initialize the model with empty weights (uninitialized memory).
+                If ``None``, the strategy will decide. Some strategies may not support all options.
+                Set this to ``True`` if you are loading a checkpoint into a large model. Requires `torch >= 1.13`.
+        """
+        if not _TORCH_GREATER_EQUAL_2_0 and self.strategy.root_device.type != "cpu":
+            rank_zero_warn(
+                "`Trainer.init_module()` can't place tensors on the device directly"
+                " with PyTorch < 2.0. Parameters will remain on CPU until the trainer starts."
+                " Upgrade to PyTorch >= 2.0 to fully utilize this feature.",
+                category=PossibleUserWarning,
+            )
+        if is_overridden("model_sharded_context", self.strategy, parent=Strategy):
```

Comment:
I think this is too often a false positive. While DeepSpeed overrides this method, it really only applies the context manager for stage 3. So for stage 1 and 2 this warning would be a false positive always. Same for FSDP with non-standard configuration. You'd need to reach into the strategies and check the configuration.
For simplicity, I recommend not doing this warning.

Reply:
When users are using `trainer.init_module` with non-sharded and switch to FSDP, how do they find out that now they need to change their code and instantiate a hook? Just by knowing this from the docs? If they don't know this detail FSDP will silently not work and that is bad

Reply:
So how would you address my comment then?

Reply:
So far the options are:
(1): Have no warning
(2): Have the warning, accept false positives
(3): Have the warning, add logic to reduce false positives
(4): don't add this method

All of them have their flaws. I would go with (2) which is what's currently implemented